BITAU-76 Show download Bitwarden action card

ahaisting-livefront opened the pull request:
## 🎟️ Tracking

https://livefront.atlassian.net/browse/BITAU-76

## 📔 Objective

The goal of this PR is to a "Download Bitwarden App" action card when the user doesn't have the main bitwarden app installed.

The card should take the user to the play store when clicked. If the user dismisses the card, they should never see it again. 

## 📸 Screenshots

<img width=300 src="https://github.com/user-attachments/assets/8521dbe1-560e-408c-9fdb-004b4abcbb30" >
<img width=300 src="https://github.com/user-attachments/assets/f15ebdfe-3cad-43d7-b504-2e0f2ea60ac2" >
<img width=300 src="https://github.com/user-attachments/assets/5065c5dc-80e8-4d46-bce3-d2aff467224a" >




## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

ahaisting-livefront commented on app/src/main/kotlin/com/bitwarden/authenticator/data/authenticator/manager/TotpCodeManagerImpl.kt:
```diff
@@ -28,6 +29,9 @@ class TotpCodeManagerImpl @Inject constructor(
     override fun getTotpCodesFlow(
         itemList: List<AuthenticatorItem>,
     ): Flow<List<VerificationCodeItem>> {
+        if (itemList.isEmpty()) {
```
This is one of two changes that is technically unrelated to this PR. 

I had a bug here where when mapping an empty list, I got a flow that never emits. This change instead maps to a flow that just emits empty list, which gets rendered correctly as the empty state.

If you'd like me to pull this into a separate PR quick, I'd be happy to!

SaintPatrck replied:
Nice catch! I think it would be good to extract this fix and include a test for it if we can.

ahaisting-livefront replied:
https://github.com/bitwarden/authenticator-android/pull/231 🤗

ahaisting-livefront commented on app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingScreen.kt:
```diff
@@ -197,20 +211,24 @@ fun ItemListingScreen(
-        ItemListingState.ViewState.Loading,
-        ItemListingState.ViewState.NoItems,
-        -> {
+        ItemListingState.ViewState.Loading -> Unit
```
This fixed a little bug where instead of loading state being shown, the empty state was shown while items loaded. 

For now I'm just showing an empty screen for loading state. It makes the transition on app startup less jarring when compared to showing empty state for a frame then full state.

ahaisting-livefront commented on app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingScreen.kt:
```diff
@@ -197,20 +211,24 @@ fun ItemListingScreen(
             )
         }
 
-        is ItemListingState.ViewState.Error -> {
-            Text(
-                text = "Error! ${currentState.message}",
-                modifier = Modifier.fillMaxSize(),
-            )
-        }
-
```
I got rid of the Error and no network states, since they were technically impossible to show. 

If you think that's going too far and it'd be nice to have them around in case we want them later, I could add them back. But for now I wanted to only worry about the possible states 👍

SaintPatrck commented on app/src/main/kotlin/com/bitwarden/authenticator/data/authenticator/repository/AuthenticatorRepositoryImpl.kt:
```diff
@@ -146,7 +144,7 @@ class AuthenticatorRepositoryImpl @Inject constructor(
     @OptIn(ExperimentalCoroutinesApi::class)
     override val sharedCodesStateFlow: StateFlow<SharedVerificationCodesState> by lazy {
-        if (!featureFlagManager.getFeatureFlag(LocalFeatureFlag.BitwardenAuthenticationEnabled)) {
+        if (!featureFlagManager.getFeatureFlag(LocalFeatureFlag.PasswordManagerSync)) {
```
Just for process consistency, let's extract this change too.

SaintPatrck commented on app/src/main/kotlin/com/bitwarden/authenticator/data/platform/datasource/disk/SettingsDiskSourceImpl.kt:
```diff
@@ -97,6 +99,12 @@ class SettingsDiskSourceImpl(
         get() = mutableIsCrashLoggingEnabledFlow
             .onSubscription { emit(getBoolean(CRASH_LOGGING_ENABLED_KEY)) }
 
+    override var hasUserDismissedDownloadBitwardenCard: Boolean?
```
Can we add tests for this?

SaintPatrck commented on app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingScreen.kt:
```diff
@@ -482,53 +527,102 @@ fun EmptyItemListingContent(
-                .padding(horizontal = 16.dp),
-            verticalArrangement = Arrangement.Center,
-            horizontalAlignment = Alignment.CenterHorizontally,
+                .padding(paddingValues = innerPadding)
```
👍

SaintPatrck commented on app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt:
```diff
@@ -403,101 +411,87 @@ class ItemListingViewModel @Inject constructor(
+            SharedVerificationCodesState.Loading,
+            SharedVerificationCodesState.OsVersionNotSupported,
+            is SharedVerificationCodesState.Success,
+                -> ItemListingState.ActionCardState.None
```
⛏️ This is odd. In PM `->` is not indented. 🤔

SaintPatrck replied:
This must be due to a recent change in the AS formatter settings. It's happening in the PM app now, too. Instead of fighting the formatter we're going to keep the indentions. I'll whip up a PR to update all the instances in one shot.

SaintPatrck reviewed (commented):
Looking good. 🚀 

Can we get test coverage on the new code? It would be nice if we can start adding coverage as we introduce new code or make changes to existing lines.

ahaisting-livefront commented:
Thanks for the feedback, @SaintPatrck !

Things I updated:
- Rebased with `main`
- Wrote tests for ItemListingViewModel,  SettingsRepository, SettingsDiskSource... I only wrote tests for the code in this patch, doing the whole thing would be a can of worms at the moment. 
- Pulled out those unrelated changes

This is ready for another look whenever you are!

SaintPatrck commented on app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt:
```diff
@@ -403,101 +411,87 @@ class ItemListingViewModel @Inject constructor(
-            updateStateWithVerificationCodeItems(
-                authenticatorData = state.data,
-                clearDialogState = true,
+                    is ItemListingState.ViewState.NoItems -> it.viewState.copy(
+                        actionCard = ItemListingState.ActionCardState.None,
```
Can we get a test for this state?

ahaisting-livefront replied:
Yep, thanks for the reminder! 5534b32446cf6fedb44e38d767186cc4ca94712e

SaintPatrck commented on app/src/main/kotlin/com/bitwarden/authenticator/ui/authenticator/feature/itemlisting/ItemListingViewModel.kt:
```diff
@@ -403,101 +411,87 @@ class ItemListingViewModel @Inject constructor(
-    private fun authenticatorNoNetworkReceive(
-        state: DataState.NoNetwork<List<VerificationCodeItem>>,
-    ) {
-        if (state.data != null) {
-            updateStateWithVerificationCodeItems(
-                authenticatorData = state.data,
-                clearDialogState = true,
+                    is ItemListingState.ViewState.NoItems -> it.viewState.copy(
+                        actionCard = ItemListingState.ActionCardState.None,
+                    )
```
Can we coalesce these conditions? Looks like they're doing the same thing.

ahaisting-livefront replied:
Unless I'm missing something, these can't be combined. The trick is that only 2/3 states have an action card (Loading does not, NoItems and Content do) so I have to get the type of the state before I can copy it it change the actionCard property.

Or said a different way -- not certain states have the `actionCard` property, so we have to `when` and only update those states.
